### PR TITLE
quantization: deprecate observer compute_dtype and replace with is_dynamic

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85431

Summary:

This PR deprecates the `compute_dtype` field on observers, and replaces
it with the `is_dynamic` field on observers.  This is better aligned
with the reference model spec.

Test plan:

```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @jgong5 @Xia-Weiwen @leslie-fang-intel @ezyang @SherlockNoMad @soumith @EikanWang @wenzhe-nrv

### CODE REVIEW

File: torch/ao/quantization/utils.py [hunk 1]
```diff
@@ -229,41 +229,42 @@ def op_is_int8_dynamically_quantized(qconfig) -> bool:
...
+    return (activation.dtype, weight.dtype, act_is_dynamic)
 
 def get_quant_type(qconfig):
     assert qconfig is not None
     activation = qconfig.activation()
     weight = qconfig.weight()
     static_dtypes = [torch.quint8, torch.qint8, torch.quint4x2]
     if weight.dtype in static_dtypes:
-        if hasattr(activation, 'compute_dtype') and activation.compute_dtype in static_dtypes:
+        # if hasattr(activation, 'compute_dtype') and activation.compute_dtype in static_dtypes:
```
Comment (jerryzh168) [inline torch/ao/quantization/utils.py:258]: remove?

**Review by jerryzh168 — APPROVED** [review]
great, thanks a lot for the clean up!

File: torch/ao/quantization/fx/convert.py [hunk 2]
```diff
@@ -1,4 +1,4 @@
-from typing import Any, Dict, List, Optional, Set, Tuple, Union, Type
+from typing import Any, Dict, List, Optional, Set, Tuple, Union, Type, Callable
```
Comment (vkuzo) [inline torch/ao/quantization/fx/convert.py:1]: There were various mypy errors in the master branch version of this file in my local env, so this PR also fixes these errors.  Technically some of this stuff could be in a separate PR, lmk if that's needed.

File: torch/ao/quantization/fx/convert.py [hunk 3]
```diff
@@ -131,11 +131,13 @@ def _replace_observer_with_quantize_dequantize_node_decomposed(
     # 1. extract the information from activation_post_process module for generating
     # the quantize and dequantize operator
     dtype = activation_post_process.dtype  # type: ignore[attr-defined]
-    compute_dtype = None
-    if hasattr(activation_post_process, "compute_dtype"):
-        compute_dtype = activation_post_process.compute_dtype  # type: ignore[attr-defined]
+
+    is_dynamic = False
+    if hasattr(activation_post_process, "is_dynamic"):
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/convert.py:136]: nit: should we make this a required attribute going forward?

File: torch/ao/quantization/fx/convert.py [hunk 4]
```diff
@@ -131,11 +131,13 @@ def _replace_observer_with_quantize_dequantize_node_decomposed(
...
+        is_dynamic = activation_post_process.is_dynamic  # type: ignore[assignment]
+
     if dtype in [torch.quint8, torch.qint8, torch.qint32] and \
-            not hasattr(activation_post_process, 'compute_dtype'):
+            (not is_dynamic):
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/convert.py:140]: nit (can be done in a separate PR): I think the overall checking can be made clearer as well for this code, right now it is:
```
if dtype in [torch.quint8, torch.qint8, torch.qint32] and not is_dynamic:
   ...
elif is_dynamic:
   ...
else:
   # fp16
```

I think a clearer one should be:
```
if is_dynamic:
   ...
else:
   if dtype in [torch.quint8, torch.qint8, torch.qint32]:
      ...
    else:
      # fp16
      ...
```

Comment (jerryzh168) [reply torch/ao/quantization/fx/convert.py:140]: also we might want to add dtype check for is_dynamic branch as well, looks like we only support uint8, int8, fp16 in the non-decomposed function and uint8 and int8 in the decomposed function

File: torch/ao/quantization/fx/utils.py [hunk 5]
```diff
@@ -190,12 +190,25 @@ def get_quantize_node_info(
...
+
+    is_dynamic = False
+    if hasattr(activation_post_process, "is_dynamic"):
+        is_dynamic = activation_post_process.is_dynamic  # type: ignore[attr-defined]
+
     quantize_op : Optional[Union[Callable, str]] = None
-    if dtype in [torch.quint8, torch.qint8, torch.qint32] and \
-            not hasattr(activation_post_process, 'compute_dtype'):
+
+    if is_dynamic:
```
Comment (jerryzh168) [inline torch/ao/quantization/fx/utils.py:200]: I think this function is removed already? could you rebase on master?

Comment (jerryzh168) [reply torch/ao/quantization/fx/utils.py:200]: oh it's not landed yet..

**Review by jerryzh168 — APPROVED** [review]
looks good, thanks a lot, please wait until https://github.com/pytorch/pytorch/pull/89272 is merged so we can remove the changes to `get_quantize_info` function, also had some optional nits that can be fixed later
